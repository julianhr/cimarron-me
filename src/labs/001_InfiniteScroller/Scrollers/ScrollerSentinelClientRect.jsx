import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'
import { throttle } from 'lodash-es'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: scroll;
`

export class ScrollerSentinelClientRect extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func.isRequired,
    isFetching: PropTypes.bool.isRequired,
    sentinelPosition: PropTypes.number.isRequired,
    setEntryCount: PropTypes.func.isRequired,
    setIsFetching: PropTypes.func.isRequired,
  }

  state = {
    cards: [],
    fetch: {
      status: 'loading',
      error: null,
    },
  }

  refRoot = React.createRef()
  refSentinel = React.createRef()

  componentDidMount() {
    this.props.setEntryCount(0)
    this.getNewCards()
  }

  handleOnScroll() {
    const { isFetching } = this.props
    const { cards } = this.state
    const { top: sentinelTop } = this.refSentinel.current.getBoundingClientRect()
    const { bottom: rootBottom } = this.refRoot.current.getBoundingClientRect()
    const isSentinelVisible = sentinelTop <= rootBottom

    if (!isFetching && isSentinelVisible && cards.length < 200) {
      this.getNewCards()
    }
  }

  async getNewCards() {
    const { cardFetcher, setIsFetching, setEntryCount } = this.props

    try {
      setIsFetching(true)
      const data = await cardFetcher(this)
      this.setNewCards(data)
      setEntryCount(this.state.cards.length)
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
    }

    setIsFetching(false)
  }

  setNewCards(data) {
    const { sentinelPosition } = this.props
    const { length: lastKey } = this.state.cards
    const cards = [...this.state.cards]

    data.forEach((datum, i) => {
      const card = this.getCard(datum, i, sentinelPosition, lastKey)
      cards.push(card)
    })

    this.setState({ cards, fetch: null })
  }

  getCard(datum, i, sentinelPosition, lastKey) {
    const { title, image_url: imgUrl, description } = datum
    const position = lastKey + i + 1
    const props = { imgUrl, title, description, position }
    let CardComponent = Card

    if (i + 1 === sentinelPosition) {
      CardComponent = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
      // previous ref element will be overwritten by new sentinel
      props.forwardedRef = this.refSentinel
    }

    return <CardComponent key={lastKey + i} {...props} />
  }

  renderResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error') {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root
        ref={this.refRoot}
        onScroll={throttle(this.handleOnScroll, 200, { leading: false }).bind(this)}
      >
        {this.renderResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching }) => (
  { sentinelPosition, isFetching }
)

const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelClientRect)
