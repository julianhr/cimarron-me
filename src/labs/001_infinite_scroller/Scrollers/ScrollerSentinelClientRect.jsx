import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: scroll;
`

class ScrollerSentinelClientRect extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func,
    isFetching: PropTypes.bool,
    sentinelPosition: PropTypes.number,
    setEntryCount: PropTypes.func,
    setIsFetching: PropTypes.func,
  }

  state = {
    cards: []
  }

  refRoot = React.createRef()
  refSentinel = React.createRef()

  componentDidMount() {
    this.props.setEntryCount(0)
    this.fetchCards()
  }

  handleOnScroll = (event) => {
    const { isFetching } = this.props
    const { cards } = this.state
    const { top: sentinelTop } = this.refSentinel.current.getBoundingClientRect()
    const { clientHeight: rootClientHeight } = this.refRoot.current
    const isSentinelVisible = sentinelTop <= rootClientHeight

    if (!isFetching && isSentinelVisible && cards.length < 200) {
      this.fetchCards()
    }
  }

  async fetchCards() {
    const { cardFetcher, setIsFetching } = this.props
    let data
    setIsFetching(true)

    try {
      data = await cardFetcher(this)
      if (!data) { return }
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
      return
    }

    const { cards: currCards } = this.state
    const cards = [...currCards]

    this.appendNewCards(cards, data, currCards.length)
    this.props.setEntryCount(cards.length)
    this.setState({ cards, fetch: null })
    setIsFetching(false)
  }

  appendNewCards(cards, data, lastKey) {
    const { sentinelPosition } = this.props

    data.forEach(({ title, image_url: imgUrl, description }, i) => {
      const position = lastKey + i + 1
      const props = { imgUrl, title, description, position }
      let CardComponent = Card

      if (i+1 === sentinelPosition) {
        CardComponent = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
        props.forwardedRef = this.refSentinel
      }

      cards.push( <CardComponent key={lastKey+i} {...props} /> )
    })
  }

  renderCardResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error') {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root
        ref={this.refRoot}
        onScroll={this.handleOnScroll}
      >
        {this.renderCardResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching }) => (
  { sentinelPosition, isFetching }
)

const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelClientRect)
