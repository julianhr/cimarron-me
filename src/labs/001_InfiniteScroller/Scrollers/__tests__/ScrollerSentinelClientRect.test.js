import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'

import { ScrollerSentinelClientRect } from '../ScrollerSentinelClientRect'


describe('ScrollerSentinelClientRect', () => {
  function fakeDiv(scrollHeight, scrollTop, clientHeight) {
    return { scrollHeight, scrollTop, clientHeight, }
  }

  const testProps = {
    cardFetcher: async () => ([]),
    isFetching: false,
    sentinelPosition: 5,
    setEntryCount: () => {},
    setIsFetching: () => {},
  }

  describe('props', () => {
    test('cardFetcher is required', () => {
      const props = {...testProps}
      delete props.cardFetcher
      expect(() => render(<ScrollerSentinelClientRect {...props} />)).toThrow()
    })

    test('isFetching is required', () => {
      const props = {...testProps}
      delete props.isFetching
      expect(() => render(<ScrollerSentinelClientRect {...props} />)).toThrow()
    })

    test('sentinelPosition is required', () => {
      const props = {...testProps}
      delete props.sentinelPosition
      expect(() => render(<ScrollerSentinelClientRect {...props} />)).toThrow()
    })

    test('setEntryCount is required', () => {
      const props = {...testProps}
      delete props.setEntryCount
      expect(() => render(<ScrollerSentinelClientRect {...props} />)).toThrow()
    })

    test('setIsFetching is required', () => {
      const props = {...testProps}
      delete props.setIsFetching
      expect(() => render(<ScrollerSentinelClientRect {...props} />)).toThrow()
    })
  })

  describe('#getCard', () => {
    const testDatum = {
      image_url: 'http://image.com',
      title: 'test title',
      description: ['sentence one', 'sentence two'],
      position: 4,
    }

    it('returns a non-sentinel Card element if datum index != sentinel index', () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      const instance = wrapper.getInstance()
      const article = document.createElement('article')
      const i = 2, sentinelPosition = 4, lastKey = 10

      instance.refSentinel = article
      expect(instance.getCard(testDatum, i, sentinelPosition, lastKey)).toMatchSnapshot()
    })

    it('returns a sentinel Card element if datum index == sentinel index', () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      const instance = wrapper.getInstance()
      const article = document.createElement('article')
      const i = 2, sentinelPosition = 3, lastKey = 10

      instance.refSentinel = article
      expect(instance.getCard(testDatum, i, sentinelPosition, lastKey)).toMatchSnapshot()
    })
  })

  describe('#setNewCards', () => {
    it('appends new cards to the current ones', () => {
      // setup env
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'getCard').mockImplementation(arg => arg)
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'getNewCards').mockReturnValue()
      const spySetState = jest.spyOn(ScrollerSentinelClientRect.prototype, 'setState')
      const testData1 = ['card 1', 'card 2']
      const testData2 = ['card 3', 'card 4']
      // render
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      const instance = wrapper.getInstance()
      // test with no data
      instance.setState({ fetch: null })
      expect(wrapper).toMatchSnapshot('no data')
      // test on data set 1
      instance.setNewCards(testData1)
      expect(spySetState).toBeCalledWith({ cards: testData1, fetch: null })
      expect(wrapper).toMatchSnapshot('data 1')
      // test on data set 2
      instance.setNewCards(testData2)
      expect(spySetState).toBeCalledWith({ cards: [...testData1, ...testData2], fetch: null })
      expect(wrapper).toMatchSnapshot('data 2')
    })
  })

  describe('#getNewCards', () => {
    let mockSetIsFetching, mockCardFetcher, mockSetNewCards, mockSetState, mockSetEntryCount
    let testCards = ['card1', 'card2', 'card3']

    beforeEach(() => {
      mockSetIsFetching = jest.spyOn(testProps, 'setIsFetching')
      mockSetEntryCount = jest.spyOn(testProps, 'setEntryCount')
      mockCardFetcher = jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => Promise.resolve(testCards))
      mockSetNewCards = jest.spyOn(ScrollerSentinelClientRect.prototype, 'setNewCards').mockReturnValue(null)
      mockSetState = jest.spyOn(ScrollerSentinelClientRect.prototype, 'setState')
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'componentDidMount').mockReturnValue(null)
    })

    it('is async', async () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      await expect(wrapper.getInstance().getNewCards()).resolves.toBe(undefined)
    })

    it('sets fetching status correctly', async () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockSetIsFetching.mock.calls).toEqual([[true], [false]])
    })
    
    it('awaits #cardFetcher', async () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockCardFetcher).toHaveBeenCalled()
    })
    
    it('calls #setNewCards', async () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockSetNewCards).toHaveBeenCalledTimes(1)
      expect(mockSetNewCards).toHaveBeenCalledWith(testCards)
    })

    it('calls #setEntryCount', async () => {
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      const instance = wrapper.getInstance()
      instance.setState({ cards: testCards })

      mockSetEntryCount.mockClear()
      await instance.getNewCards()
      expect(mockSetEntryCount).toHaveBeenCalledTimes(1)
      expect(mockSetEntryCount).toHaveBeenCalledWith(testCards.length)
    })

    it('catches error if #cardFetcher fails', async () => {
      mockCardFetcher.mockRestore()
      mockCardFetcher = jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => Promise.reject('mocked error'))
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      await wrapper.getInstance().getNewCards()
      expect(mockSetState).toHaveBeenCalled()
    })
  })

  describe('#handleOnScroll', () => {
    let propsHandler, mockCanFetchCards, mockGetNewCards

    beforeEach(() => {
      propsHandler = {...testProps}
      propsHandler.isFetching = false
      mockGetNewCards = jest.spyOn(ScrollerSentinelClientRect.prototype, 'getNewCards').mockReturnValue(null)
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'componentDidMount').mockReturnValue(null)
    })

    it('should call #getNewCards when all conditions match', () => {
      const wrapper = create(<ScrollerSentinelClientRect {...propsHandler} />)
      const instance = wrapper.getInstance()
      instance.refRoot.current = { getBoundingClientRect: () => ({ bottom: 200 }) }
      instance.refSentinel.current = { getBoundingClientRect: () => ({ top: 199 }) }
      instance.handleOnScroll()
      expect(mockGetNewCards).toHaveBeenCalledTimes(1)
    })

    describe('should not call #getNewCards if', () => {
      test('instance is fetching', () => {
        propsHandler.isFetching = true
        const wrapper = create(<ScrollerSentinelClientRect {...propsHandler} />)
        const instance = wrapper.getInstance()
        instance.refRoot.current = { getBoundingClientRect: () => ({ bottom: 200 }) }
        instance.refSentinel.current = { getBoundingClientRect: () => ({ top: 199 }) }
        instance.handleOnScroll()
        expect(mockGetNewCards).not.toHaveBeenCalled()
      })

      test('sentinel is not visible', () => {
        const wrapper = create(<ScrollerSentinelClientRect {...propsHandler} />)
        const instance = wrapper.getInstance()
        instance.refRoot.current = { getBoundingClientRect: () => ({ bottom: 200 }) }
        instance.refSentinel.current = { getBoundingClientRect: () => ({ top: 201 }) }
        instance.handleOnScroll()
        expect(mockGetNewCards).not.toHaveBeenCalled()
      })

      test('card.legth >= 200', () => {
        // env setup
        const wrapper = create(<ScrollerSentinelClientRect {...propsHandler} />)
        const instance = wrapper.getInstance()
        instance.refRoot.current = { getBoundingClientRect: () => ({ bottom: 200 }) }
        instance.refSentinel.current = { getBoundingClientRect: () => ({ top: 199 }) }
        // cards
        const mockEntry = 'card'
        let cards = new Array(199).fill(mockEntry)

        // 199 entries
        instance.setState({ cards })
        instance.handleOnScroll()
        expect(mockGetNewCards).toHaveBeenCalledTimes(1)
        // 200 entries
        cards = [mockEntry, ...cards]
        instance.setState({ cards })
        instance.handleOnScroll()
        expect(mockGetNewCards).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('Intergration', () => {
    it('matches snapshot for happy path', async () => {
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'getNewCards').mockReturnValue([])
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      const cards = [<div>one</div>, <div>two</div>]
      wrapper.getInstance().setState({ fetch: null, cards })
      expect(wrapper).toMatchSnapshot()
    })

    it('displays loading message on mount', () => {
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'getNewCards').mockReturnValue(['one', 'two'])
      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('displays error message on mount if initial fetch throws error', async () => {
      jest.spyOn(testProps, 'cardFetcher')
        .mockImplementation(() => { throw new Error('mocked error') })

      const wrapper = create(<ScrollerSentinelClientRect {...testProps} />)
      expect(wrapper).toMatchSnapshot()
    })
    
    it('displays cards up to point if error thrown by #cardFetcher', async () => {
      const cards = ['one', 'two', 'three']
      jest.spyOn(ScrollerSentinelClientRect.prototype, 'getNewCards')
        .mockImplementation(function() { this.setState({ cards, fetch: null }) })
      const wrapper = mount(<ScrollerSentinelClientRect {...testProps} />)
      const instance = wrapper.instance()

      jest.restoreAllMocks()
      jest.spyOn(testProps, 'cardFetcher').mockImplementation(() => { throw new Error('mocked error') })
      expect(wrapper).toMatchSnapshot()
      
      const html = wrapper.html()
      instance.getNewCards()
      wrapper.update()
      expect(wrapper.html()).toEqual(html)
      wrapper.unmount()
    })
  })
})
