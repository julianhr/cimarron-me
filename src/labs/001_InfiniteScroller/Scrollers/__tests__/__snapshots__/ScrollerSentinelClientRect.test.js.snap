// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollerSentinelClientRect #getCard returns a non-sentinel Card element if datum index != sentinel index 1`] = `
<Card
  description={
    Array [
      "sentence one",
      "sentence two",
    ]
  }
  imgUrl="http://image.com"
  position={13}
  title="test title"
/>
`;

exports[`ScrollerSentinelClientRect #getCard returns a sentinel Card element if datum index == sentinel index 1`] = `
<ForwardRef
  description={
    Array [
      "sentence one",
      "sentence two",
    ]
  }
  forwardedRef={<article />}
  imgUrl="http://image.com"
  position={13}
  title="test title"
/>
`;

exports[`ScrollerSentinelClientRect #setNewCards appends new cards to the current ones: data 1 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
>
  card 1
  card 2
</div>
`;

exports[`ScrollerSentinelClientRect #setNewCards appends new cards to the current ones: data 2 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
>
  card 1
  card 2
  card 3
  card 4
</div>
`;

exports[`ScrollerSentinelClientRect #setNewCards appends new cards to the current ones: no data 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
/>
`;

exports[`ScrollerSentinelClientRect Intergration displays cards up to point if error thrown by #cardFetcher 1`] = `
<Root
  onScroll={[Function]}
>
  <div
    className="css-1ml1922-Root e1u23dax0"
    onScroll={[Function]}
  >
    one
    two
    three
  </div>
</Root>
`;

exports[`ScrollerSentinelClientRect Intergration displays error message on mount if initial fetch throws error 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
>
  <div
    className="css-quardp-Message ehwlun20"
  >
    <div>
      Error: mocked error
    </div>
  </div>
</div>
`;

exports[`ScrollerSentinelClientRect Intergration displays loading message on mount 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
>
  <div
    className="css-quardp-Message ehwlun20"
  >
    <div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`ScrollerSentinelClientRect Intergration matches snapshot for happy path 1`] = `
<div
  className="css-1ml1922-Root e1u23dax0"
  onScroll={[Function]}
>
  <div>
    one
  </div>
  <div>
    two
  </div>
</div>
`;
